File: sdks/node/src/meta-proxy/routing.ts
/**
 * Data-plane routing intent and decode-time verification (ADR-0025a §D5).
 *
 * This module carries the caller's *supported intent* and verifies the
 * Proxy's decision against it — it deliberately does NOT implement a router.
 * ADR-0025a §D5: "The SDK communicates supported intent and verifies the
 * decision; it does not implement another router." There is no planner, no
 * plane selector, and no failover logic here; the Proxy owns all of that.
 *
 * `RoutingPlane` / `WorkloadPolicy` are re-exported from `./status.js` (where
 * §D4 already needed them for `MetaProxyStatus`) rather than redeclared, so
 * the union types have exactly one definition across the module.
 */

import { AgenticError } from "../agentic/index.js";
import type { MetaProxyRoutingReceipt, RoutingPlane, WorkloadPolicy } from "./status.js";

export type { RoutingPlane, WorkloadPolicy } from "./status.js";

const PRODUCT = "meta-proxy";

/**
 * A consent grant is an opaque ADR-0022 grant identifier — the SDK treats it
 * as a bare string ID and does not interpret its structure (ADR-0025a §D9
 * owns consent semantics; this pass only forwards intent).
 */
export type ConsentGrantId = string;

/**
 * Supported routing intent a caller attaches to an inference call
 * (ADR-0025a §D5). The Proxy is the authority; the SDK sends this as intent
 * and verifies the returned receipt against it (see
 * {@link assertRoutingReceiptMatchesIntent}). None of these fields cause the
 * SDK to *choose* a plane.
 */
export interface RoutingIntent {
  /**
   * If set, the returned receipt's `selectedPlane` MUST equal this or the
   * call is a protocol violation "even if output succeeds" (ADR-0025a §D5
   * rule 7). Verified at decode time by
   * {@link assertRoutingReceiptMatchesIntent}.
   */
  requiredPlane?: RoutingPlane;
  /** Planes the caller will accept. Advisory intent — the Proxy enforces. */
  allowedPlanes: RoutingPlane[];
  /** Workload urgency class (ADR-0025a §D5). `critical` suppresses automatic failover. */
  workloadPolicy: WorkloadPolicy;
  /** Ceiling the caller is willing to route under, when the Proxy exposes utilization. */
  maxUtilization?: number;
  /** Opaque ADR-0022 consent grant IDs relevant to this call (ADR-0025a §D5/§D9). */
  consentGrants: ConsentGrantId[];
  /** Whether the caller opts into training contribution (reported without content, §D9). */
  trainingShare: boolean;
  /** If true, the Proxy must fail rather than silently degrade to another plane (§D5 rule 5). */
  failIfUnavailable: boolean;
}

/**
 * The single decode-time verification §D5 mandates (rule 7): a
 * `requiredPlane` that the returned receipt contradicts is a protocol
 * violation, non-retryable, and MUST throw even when the HTTP call itself
 * was a well-formed 200. This is the SDK's *only* routing "decision" — a
 * pure after-the-fact check, never a selection.
 *
 * Throws {@link AgenticError} `kind: "protocol"` when:
 *  - `intent.requiredPlane` is set and no receipt was returned to verify
 *    against (§D4: "Every inference must return selected-plane evidence"), or
 *  - the receipt's `selectedPlane` does not equal `intent.requiredPlane`.
 *
 * A no-op when `intent` is undefined or carries no `requiredPlane`.
 */
export function assertRoutingReceiptMatchesIntent(
  intent: RoutingIntent | undefined,
  receipt: MetaProxyRoutingReceipt | undefined,
): void {
  const required = intent?.requiredPlane;
  if (!required) return;

  if (!receipt) {
    throw new AgenticError(
      "protocol",
      `RoutingIntent.requiredPlane "${required}" cannot be verified: the Proxy ` +
        `response carried no routing receipt (ADR-0025a §D4: every inference must ` +
        `return selected-plane evidence)`,
      {
        product: PRODUCT,
        operation: "chat.completions",
        retryable: false,
        details: { requiredPlane: required },
      },
    );
  }

  if (receipt.selectedPlane !== required) {
    throw new AgenticError(
      "protocol",
      `Proxy selected plane "${receipt.selectedPlane}" but RoutingIntent.requiredPlane ` +
        `was "${required}" — a required-plane mismatch is a protocol violation even ` +
        `when output succeeds (ADR-0025a §D5 rule 7)`,
      {
        product: PRODUCT,
        operation: "chat.completions",
        requestId: receipt.requestId,
        retryable: false,
        details: {
          requiredPlane: required,
          selectedPlane: receipt.selectedPlane,
          configuredPlane: receipt.configuredPlane,
        },
      },
    );
  }
}

File: sdks/python/cognitum/meta_proxy/routing.py
"""Data-plane routing intent and decode-time receipt verification (ADR-0025a §D5).

The SDK communicates supported intent and verifies the Proxy's decision; it
does NOT implement another router (§D5: "it does not implement another
router"). There is deliberately no planner, selector, or plane-choosing
logic in this module -- only a description of what the caller is willing to
accept (:class:`RoutingIntent`) plus the single decode-time check §D5 rule 7
mandates (:func:`assert_routing_receipt_matches_intent`).

``RoutingPlane`` and ``WorkloadPolicy`` are re-exported from
:mod:`cognitum.meta_proxy.status` -- they are declared there (referenced by
``MetaProxyStatus``) and are NOT redefined here.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from cognitum.agentic import AgenticError
from cognitum.meta_proxy.status import (
    MetaProxyRoutingReceipt,
    RoutingPlane,
    WorkloadPolicy,
)

if TYPE_CHECKING:
    pass

_PRODUCT = "meta-proxy"


@dataclass(frozen=True)
class RoutingIntent:
    """The plane/policy envelope a caller is willing to accept for one
    inference call (ADR-0025a §D5).

    This is a declaration, not an instruction: the Proxy owns the actual
    routing decision. The SDK only (a) communicates this intent and (b)
    verifies the returned receipt against it
    (:func:`assert_routing_receipt_matches_intent`).

    ``workload_policy`` defaults to ``"standard"`` -- the neutral policy in
    §D5's ``critical | standard | economy`` set (``critical`` suppresses
    automatic failover per rule 3, so it is never a safe silent default).
    """

    #: If set, the receipt's ``selected_plane`` MUST equal this or the call
    #: is a protocol violation "even if output succeeds" (§D5 rule 7).
    required_plane: RoutingPlane | None = None
    #: Planes the caller will accept. Empty means "no explicit restriction
    #: beyond ``required_plane``"; the SDK does not enforce this list itself
    #: (no SDK router, §D5) -- it is transmitted for the Proxy to honor.
    allowed_planes: list[RoutingPlane] = field(default_factory=list)
    workload_policy: WorkloadPolicy = "standard"
    #: Utilization ceiling above which the caller prefers not to route to an
    #: automatic plane. Implementation data, not an SDK constant (§D5).
    max_utilization: float | None = None
    #: ADR-0022 consent grant IDs the caller is presenting. Credential
    #: presence is not consent (§D9); these are the explicit grants.
    consent_grants: list[str] = field(default_factory=list)
    #: Whether the caller opts into training contribution. Independent of
    #: routing and reported without content (§D9).
    training_share: bool = False
    #: Whether an unavailable required/allowed plane should fail rather than
    #: silently degrade. Pairs with §D5 rule 5 ("local failure does not
    #: authorize cloud egress").
    fail_if_unavailable: bool = True


def assert_routing_receipt_matches_intent(
    intent: RoutingIntent | None,
    receipt: MetaProxyRoutingReceipt,
) -> None:
    """Decode-time verification of §D5 rule 7: a ``required_plane`` mismatch
    is a protocol violation "even if output succeeds".

    Raises a non-retryable ``AgenticError("protocol", ...)`` when
    ``intent.required_plane`` is set and does not equal
    ``receipt.selected_plane`` -- this MUST fire even when the HTTP call was
    a well-formed 200, because the SDK "never infers plane" and treats a
    receipt contradicting caller intent as a protocol error (§D4/§D5).

    A ``None`` intent, or an intent with no ``required_plane``, is a no-op:
    the SDK does not run its own router, so it has nothing to assert beyond
    the caller's explicit ``required_plane`` demand.
    """
    if intent is None or intent.required_plane is None:
        return
    if receipt.selected_plane != intent.required_plane:
        raise AgenticError(
            "protocol",
            "Meta Proxy routing receipt contradicts caller intent: "
            f'required_plane "{intent.required_plane}" but the Proxy reported '
            f'selected_plane "{receipt.selected_plane}" (ADR-0025a §D5 rule 7: a '
            "required_plane mismatch is a protocol violation even if output "
            "succeeds)",
            product=_PRODUCT,
            operation="chat.completions",
            request_id=receipt.request_id or None,
            retryable=False,
        )


__all__ = [
    "RoutingPlane",
    "WorkloadPolicy",
    "RoutingIntent",
    "assert_routing_receipt_matches_intent",
]
